# Практическая работа

# Задача

Один из способов повысить эффективность взаимодействия банка с клиентами — отправлять предложение о новой услуге не всем клиентам, а только некоторым, которые выбираются по принципу наибольшей склонности к отклику на это предложение.

Задача заключается в том, чтобы предложить алгоритм, который будет выдавать склонность клиента к положительному или отрицательному отклику на предложение банка. Предполагается, что, получив такие оценки для некоторого множества клиентов, банк обратится с предложением только к тем, от кого ожидается положительный отклик.


Для решения этой задачи загрузите файлы из базы в Postgres.
Эта БД хранит информацию о клиентах банка и их персональные данные, такие как пол, количество детей и другие.

Описание таблиц с данными представлено ниже.


**D_work**

Описание статусов относительно работы:
- ID — идентификатор социального статуса клиента относительно работы;
- COMMENT — расшифровка статуса.


**D_pens**

Описание статусов относительно пенсии:
- ID — идентификатор социального статуса;
- COMMENT — расшифровка статуса.


**D_clients**

Описание данных клиентов:
- ID — идентификатор записи;
- AGE	— возраст клиента;
- GENDER — пол клиента (1 — мужчина, 0 — женщина);
- EDUCATION — образование;
- MARITAL_STATUS — семейное положение;
- CHILD_TOTAL	— количество детей клиента;
- DEPENDANTS — количество иждивенцев клиента;
- SOCSTATUS_WORK_FL	— социальный статус клиента относительно работы (1 — работает, 0 — не работает);
- SOCSTATUS_PENS_FL	— социальный статус клиента относительно пенсии (1 — пенсионер, 0 — не пенсионер);
- REG_ADDRESS_PROVINCE — область регистрации клиента;
- FACT_ADDRESS_PROVINCE — область фактического пребывания клиента;
- POSTAL_ADDRESS_PROVINCE — почтовый адрес области;
- FL_PRESENCE_FL — наличие в собственности квартиры (1 — есть, 0 — нет);
- OWN_AUTO — количество автомобилей в собственности.


**D_agreement**

Таблица с зафиксированными откликами клиентов на предложения банка:
- AGREEMENT_RK — уникальный идентификатор объекта в выборке;
- ID_CLIENT — идентификатор клиента;
- TARGET — целевая переменная: отклик на маркетинговую кампанию (1 — отклик был зарегистрирован, 0 — отклика не было).
    
    
**D_job**

Описание информации о работе клиентов:
- GEN_INDUSTRY — отрасль работы клиента;
- GEN_TITLE — должность;
- JOB_DIR — направление деятельности внутри компании;
- WORK_TIME — время работы на текущем месте (в месяцах);
- ID_CLIENT — идентификатор клиента.


**D_salary**

Описание информации о заработной плате клиентов:
- ID_CLIENT — идентификатор клиента;
- FAMILY_INCOME — семейный доход (несколько категорий);
- PERSONAL_INCOME — личный доход клиента (в рублях).


**D_last_credit**

Информация о последнем займе клиента:
- ID_CLIENT — идентификатор клиента;
- CREDIT — сумма последнего кредита клиента (в рублях);
- TERM — срок кредита;
- FST_PAYMENT — первоначальный взнос (в рублях).


**D_loan**

Информация о кредитной истории клиента:
- ID_CLIENT — идентификатор клиента;
- ID_LOAN — идентификатор кредита.

**D_close_loan**

Информация о статусах кредита (ссуд):
- ID_LOAN — идентификатор кредита;
- CLOSED_FL — текущий статус кредита (1 — закрыт, 0 — не закрыт).

Ниже представлен минимальный список колонок, которые должны находиться в итоговом датасете после склейки и агрегации данных. По своему усмотрению вы можете добавить дополнительные к этим колонки.

    - AGREEMENT_RK — уникальный идентификатор объекта в выборке;
    - TARGET — целевая переменная: отклик на маркетинговую кампанию (1 — отклик был зарегистрирован, 0 — отклика не было);
    - AGE — возраст клиента;
    - SOCSTATUS_WORK_FL — социальный статус клиента относительно работы (1 — работает, 0 — не работает);
    - SOCSTATUS_PENS_FL — социальный статус клиента относительно пенсии (1 — пенсионер, 0 — не пенсионер);
    - GENDER — пол клиента (1 — мужчина, 0 — женщина);
    - CHILD_TOTAL — количество детей клиента;
    - DEPENDANTS — количество иждивенцев клиента;
    - PERSONAL_INCOME — личный доход клиента (в рублях);
    - LOAN_NUM_TOTAL — количество ссуд клиента;
    - LOAN_NUM_CLOSED — количество погашенных ссуд клиента.


Будьте внимательны при сборке датасета: это реальные банковские данные, в которых могут наблюдаться дубли, некорректно заполненные значения или значения, противоречащие друг другу. Для получения качественной модели необходимо предварительно очистить датасет от такой информации.

## Задание 1

В предыдущем задании вы собрали всю информацию о клиентах в одну таблицу, где одна строчка соответствует полной информации об одном клиенте.

Загрузите эту таблицу.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
!ls /content/drive

MyDrive


In [3]:
%cd /content/drive/MyDrive/Colab Notebooks/Линейные модели и их презентация/Project

/content/drive/MyDrive/Colab Notebooks/Линейные модели и их презентация/Project


In [4]:
import pandas as pd


df = pd.read_csv("clients_clean.csv")

In [5]:
df.head()

WORK_TIME           FAMILY_INCOME  PERSONAL_INCOME    CREDIT  TERM  \
0       18.0  от 10000 до 20000 руб.           5000.0   8000.00     6   
1       97.0  от 10000 до 20000 руб.          12000.0  21650.00     6   
2       84.0  от 10000 до 20000 руб.           9000.0  33126.00    12   
3      168.0  от 20000 до 50000 руб.          25000.0   8491.82     6   
4      101.0  от 10000 до 20000 руб.          12000.0  21990.00    12   

   FST_PAYMENT  AGREEMENT_RK  TARGET  AGE  GENDER            EDUCATION  \
0       8650.0      59910150       0   49       1  Среднее специальное   
1       4000.0      59910230       0   32       1              Среднее   
2       4000.0      59910525       0   52       1     Неполное среднее   
3       5000.0      59910803       0   39       1               Высшее   
4       4000.0      59911781       0   30       0              Среднее   

   MARITAL_STATUS  CHILD_TOTAL  DEPENDANTS  SOCSTATUS_WORK_FL  \
0  Состою в браке            2           1                  1   
1  Состою в браке            3           3                  1   
2  Состою в браке            4           0                  1   
3  Состою в браке            1           1                  1   
4  Состою в браке            0           0                  1   

   SOCSTATUS_PENS_FL  FL_PRESENCE_FL  OWN_AUTO  LOAN_NUM_TOTAL  \
0                  0               0         0               1   
1                  0               0         0               1   
2                  0               0         0               2   
3                  0               1         0               1   
4                  0               0         0               2   

   LOAN_NUM_CLOSED  
0                1  
1                1  
2                1  
3                1  
4                1

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13855 entries, 0 to 13854
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   WORK_TIME          13855 non-null  float64
 1   FAMILY_INCOME      13855 non-null  object 
 2   PERSONAL_INCOME    13855 non-null  float64
 3   CREDIT             13855 non-null  float64
 4   TERM               13855 non-null  int64  
 5   FST_PAYMENT        13855 non-null  float64
 6   AGREEMENT_RK       13855 non-null  int64  
 7   TARGET             13855 non-null  int64  
 8   AGE                13855 non-null  int64  
 9   GENDER             13855 non-null  int64  
 10  EDUCATION          13855 non-null  object 
 11  MARITAL_STATUS     13855 non-null  object 
 12  CHILD_TOTAL        13855 non-null  int64  
 13  DEPENDANTS         13855 non-null  int64  
 14  SOCSTATUS_WORK_FL  13855 non-null  int64  
 15  SOCSTATUS_PENS_FL  13855 non-null  int64  
 16  FL_PRESENCE_FL     138

Разбейте данные на тренировочную и тестовую часть в пропорции 80% к 20%, зафиксируйте `random_state = 42`.

In [7]:
X = df.drop(['TARGET', 'AGREEMENT_RK'], axis=1)
y = df['TARGET']

In [8]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [9]:
from sklearn.preprocessing import OneHotEncoder, MinMaxScaler
from sklearn.compose import ColumnTransformer

categorical = ['FAMILY_INCOME', 'EDUCATION', 'MARITAL_STATUS']
numeric_features = ['WORK_TIME', 'PERSONAL_INCOME', 'CREDIT', "TERM", "FST_PAYMENT",
                    "AGE", "GENDER", "CHILD_TOTAL", "DEPENDANTS", "SOCSTATUS_WORK_FL",
                    "SOCSTATUS_PENS_FL", "OWN_AUTO", "LOAN_NUM_TOTAL", "LOAN_NUM_CLOSED"]

ct = ColumnTransformer([
    ('ohe', OneHotEncoder(handle_unknown="ignore"), categorical),
    ('scaling', MinMaxScaler(), numeric_features)
])

X_train_transformed = ct.fit_transform(X_train)
X_test_transformed = ct.transform(X_test)

In [10]:
new_features = list(ct.named_transformers_['ohe'].get_feature_names_out())
new_features.extend(numeric_features)

new_features

['FAMILY_INCOME_до 5000 руб.',
 'FAMILY_INCOME_от 10000 до 20000 руб.',
 'FAMILY_INCOME_от 20000 до 50000 руб.',
 'FAMILY_INCOME_от 5000 до 10000 руб.',
 'FAMILY_INCOME_свыше 50000 руб.',
 'EDUCATION_Высшее',
 'EDUCATION_Два и более высших образования',
 'EDUCATION_Неоконченное высшее',
 'EDUCATION_Неполное среднее',
 'EDUCATION_Среднее',
 'EDUCATION_Среднее специальное',
 'EDUCATION_Ученая степень',
 'MARITAL_STATUS_Вдовец/Вдова',
 'MARITAL_STATUS_Гражданский брак',
 'MARITAL_STATUS_Не состоял в браке',
 'MARITAL_STATUS_Разведен(а)',
 'MARITAL_STATUS_Состою в браке',
 'WORK_TIME',
 'PERSONAL_INCOME',
 'CREDIT',
 'TERM',
 'FST_PAYMENT',
 'AGE',
 'GENDER',
 'CHILD_TOTAL',
 'DEPENDANTS',
 'SOCSTATUS_WORK_FL',
 'SOCSTATUS_PENS_FL',
 'OWN_AUTO',
 'LOAN_NUM_TOTAL',
 'LOAN_NUM_CLOSED']

In [11]:
X_train_transformed = pd.DataFrame(X_train_transformed, columns=new_features)
X_test_transformed = pd.DataFrame(X_test_transformed, columns=new_features)

X_train_transformed.head()

FAMILY_INCOME_до 5000 руб.  FAMILY_INCOME_от 10000 до 20000 руб.  \
0                         0.0                                   1.0   
1                         0.0                                   1.0   
2                         0.0                                   1.0   
3                         0.0                                   1.0   
4                         0.0                                   1.0   

   FAMILY_INCOME_от 20000 до 50000 руб.  FAMILY_INCOME_от 5000 до 10000 руб.  \
0                                   0.0                                  0.0   
1                                   0.0                                  0.0   
2                                   0.0                                  0.0   
3                                   0.0                                  0.0   
4                                   0.0                                  0.0   

   FAMILY_INCOME_свыше 50000 руб.  EDUCATION_Высшее  \
0                             0.0               0.0   
1                             0.0               0.0   
2                             0.0               0.0   
3                             0.0               0.0   
4                             0.0               0.0   

   EDUCATION_Два и более высших образования  EDUCATION_Неоконченное высшее  \
0                                       0.0                            0.0   
1                                       0.0                            0.0   
2                                       0.0                            0.0   
3                                       0.0                            0.0   
4                                       0.0                            0.0   

   EDUCATION_Неполное среднее  EDUCATION_Среднее  ...  FST_PAYMENT       AGE  \
0                         0.0                1.0  ...     0.009321  0.413043   
1                         0.0                1.0  ...     0.007143  0.239130   
2                         0.0                0.0  ...     0.013786  0.086957   
3                         0.0                1.0  ...     0.014286  0.413043   
4                         0.0                1.0  ...     0.031021  0.739130   

   GENDER  CHILD_TOTAL  DEPENDANTS  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  \
0     0.0          0.0    0.000000                1.0                0.0   
1     0.0          0.2    0.333333                1.0                0.0   
2     1.0          0.0    0.000000                1.0                0.0   
3     0.0          0.3    0.500000                1.0                0.0   
4     1.0          0.0    0.000000                1.0                1.0   

   OWN_AUTO  LOAN_NUM_TOTAL  LOAN_NUM_CLOSED  
0       1.0        0.142857             0.25  
1       0.0        0.142857             0.25  
2       0.0        0.000000             0.00  
3       0.0        0.000000             0.00  
4       0.0        0.142857             0.25  

[5 rows x 31 columns]

На тренировочных данных обучите линейную модель классификации для предсказания целевой переменной (столбец `TARGET`).

Сделайте прогноз вероятности отклика на рекламную кампанию для тестовых данных.

In [12]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_train_transformed, y_train)
pred_test = model.predict(X_test_transformed) # предсказание классов

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Переведите вероятности в классы по стандартному порогу (0.5) и на тестовом наборе данных вычислите метрики:

* accuracy
* precision
* recall
* f1-score

In [13]:
pred_test[:25]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0])

In [14]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

accuracy_score(y_test, pred_test)

0.874413569108625

In [15]:
y.value_counts(normalize=True)

0    0.872826
1    0.127174
Name: TARGET, dtype: float64

Модель, кажется, просто предсказывает всё нулями, так как метрика accuracy похожа на дисбаланс классов.

In [16]:
confusion_matrix(y_test, pred_test)

array([[2423,    1],
       [ 347,    0]])

Так и есть: для 2424 клиентов, которые не откликнулись, модель ошибочно предсказала только про 1. А вот про 347 клиентов, которые откликнулись, она всех записала в неоткликнувшихся.

In [17]:
probs = model.predict_proba(X_test_transformed)

probs[:5]

array([[0.8457627 , 0.1542373 ],
       [0.88758328, 0.11241672],
       [0.90046326, 0.09953674],
       [0.84039824, 0.15960176],
       [0.74056631, 0.25943369]])

In [18]:
probs_churn = probs[:,1]

classes = probs_churn > 0.5

In [19]:
print(f"Accuracy: {accuracy_score(y_test, classes)}")
print(f"Precision: {precision_score(y_test, classes)}")
print(f"Recall: {recall_score(y_test, classes)}")
print(f"f1: {f1_score(y_test, classes)}")

Accuracy: 0.874413569108625
Precision: 0.0
Recall: 0.0
f1: 0.0


In [20]:
confusion_matrix(y_test, classes)

array([[2423,    1],
       [ 347,    0]])

Целевая метрика для задачи - полнота, так как нам нужно найти максимум клиентов, кто может откликнуться на рекламу.

Но при этом точность не должна просесть, поэтому за ней тоже следим.

Разбейте тренировочные данные на `train` и `val` части в пропорции 3 к 1.

В цикле:

* переберите пороги от 0 до 1 с шагом 0.01
* вычислите для каждого порога значение метрик precision и recall
* подберите такой порог, при котором recall не меньше 0.66, а точность максимальна.

In [21]:
df = pd.read_csv("clients_clean.csv")

X = df.drop(['TARGET', 'AGREEMENT_RK'], axis=1)
y = df['TARGET']

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.25, random_state=42)

categorical = ['FAMILY_INCOME', 'EDUCATION', 'MARITAL_STATUS']
numeric_features = ['WORK_TIME', 'PERSONAL_INCOME', 'CREDIT', "TERM", "FST_PAYMENT",
                    "AGE", "GENDER", "CHILD_TOTAL", "DEPENDANTS", "SOCSTATUS_WORK_FL",
                    "SOCSTATUS_PENS_FL", "OWN_AUTO", "LOAN_NUM_TOTAL", "LOAN_NUM_CLOSED"]

ct = ColumnTransformer([
    ('ohe', OneHotEncoder(handle_unknown="ignore"), categorical),
    ('scaling', MinMaxScaler(), numeric_features)
])

X_train_transformed = ct.fit_transform(X_train)
X_val_transformed = ct.transform(X_val)

new_features = list(ct.named_transformers_['ohe'].get_feature_names_out())
new_features.extend(numeric_features)

X_train_transformed = pd.DataFrame(X_train_transformed, columns=new_features)
X_val_transformed = pd.DataFrame(X_val_transformed, columns=new_features)



In [22]:
precision_max = -1
limit = 0
i = 0

model = LogisticRegression()
model.fit(X_train_transformed, y_train)
pred_test = model.predict(X_val_transformed)

probs = model.predict_proba(X_val_transformed)
probs_churn = probs[:,1]

while i <= 1:
    classes = probs_churn > i
    if (recall_score(y_val, classes) >= 0.66) and (precision_score(y_val, classes) > precision_max):
        limit = i
        precision_max = precision_score(y_val, classes)
    i += 0.01

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [23]:
limit, precision_max

(0.10999999999999999, 0.1527708437343984)

Для выбранного порога посчитайте все метрики на тестовых данных. Сильно ли они отличаются от метрик на валидации?

In [24]:
classes = probs_churn > limit
print(f"Accuracy: {accuracy_score(y_val, classes)}")
print(f"Precision: {precision_score(y_val, classes)}")
print(f"Recall: {recall_score(y_val, classes)}")
print(f"f1: {f1_score(y_val, classes)}")

Accuracy: 0.47286374133949194
Precision: 0.1527708437343984
Recall: 0.7034482758620689
f1: 0.25102543068088595


In [25]:
confusion_matrix(y_val, classes)

array([[1332, 1697],
       [ 129,  306]])

Все метрики сильно отличаются. Результат стал намного лучше, потому что модель уже выявляет 306 человек, которые откликнутся.

Выведите на экран в виде таблицы топ-6 признаков с наибольшими по модулю весами модели.

In [26]:
weights = sorted(model.coef_[0], key=abs, reverse=True)
weights = weights[:6]
weights

[-2.991748842784804,
 2.715108778788992,
 -2.13033184780119,
 -1.3496723108876758,
 1.2661580138043398,
 -1.0796990304539977]

In [27]:
coef_table = pd.DataFrame({'features' : list(X_train_transformed.columns), 'weights' : list(model.coef_[0])})

coef_table.sort_values(by='weights')

features   weights
21                               FST_PAYMENT -2.991749
30                           LOAN_NUM_CLOSED -2.130332
26                         SOCSTATUS_WORK_FL -1.349672
22                                       AGE -1.079699
3        FAMILY_INCOME_от 5000 до 10000 руб. -0.339692
27                         SOCSTATUS_PENS_FL -0.317971
16             MARITAL_STATUS_Состою в браке -0.236519
5                           EDUCATION_Высшее -0.235426
1       FAMILY_INCOME_от 10000 до 20000 руб. -0.233709
11                  EDUCATION_Ученая степень -0.150070
2       FAMILY_INCOME_от 20000 до 50000 руб. -0.128843
17                                 WORK_TIME -0.115142
10             EDUCATION_Среднее специальное -0.086936
14         MARITAL_STATUS_Не состоял в браке -0.019301
6   EDUCATION_Два и более высших образования -0.009503
23                                    GENDER -0.005199
13           MARITAL_STATUS_Гражданский брак  0.081696
9                          EDUCATION_Среднее  0.088694
28                                  OWN_AUTO  0.107052
12               MARITAL_STATUS_Вдовец/Вдова  0.115317
15                MARITAL_STATUS_Разведен(а)  0.115557
25                                DEPENDANTS  0.175188
7              EDUCATION_Неоконченное высшее  0.209102
0                 FAMILY_INCOME_до 5000 руб.  0.223877
19                                    CREDIT  0.238061
8                 EDUCATION_Неполное среднее  0.240888
4             FAMILY_INCOME_свыше 50000 руб.  0.535116
29                            LOAN_NUM_TOTAL  0.706102
20                                      TERM  0.883517
24                               CHILD_TOTAL  1.266158
18                           PERSONAL_INCOME  2.715109

In [29]:
for i in range(31):
    if coef_table.loc[i]["weights"] not in weights: # смотрим по индексам
        coef_table.drop([i], inplace=True)

In [31]:
coef_table

features   weights
18    PERSONAL_INCOME  2.715109
21        FST_PAYMENT -2.991749
22                AGE -1.079699
24        CHILD_TOTAL  1.266158
26  SOCSTATUS_WORK_FL -1.349672
30    LOAN_NUM_CLOSED -2.130332

## Задание 2

Добавьте в Streamlit-приложение визуализацию результатов модели:

* опцию выбора порога и вывод метрик качества в зависимости от выбранного порога

* вывод прогноза модели на выбранном объекте (клиенте) - вероятность отклика на рекламу.

In [32]:
import pickle

with open('model.pickle', 'wb') as f:
    pickle.dump(model, f)

## Бонус

Попробуйте применить другие модели классификации для решения этой задачи (любые какие знаете).

Удалось ли добиться улучшения качества модели?

In [30]:
# your code here